### PR TITLE
external/libsodium: add sodium lib

### PR DESCRIPTION
**Pull request by SRCN-LiQin**
Add libsodium for st-device-sdk on tizenrt

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [c35e2b7f7e6ccbb4489d7d6d46dfe274dad42128] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [c35e2b7f7e6ccbb4489d7d6d46dfe274dad42128] - Code Rule Check OK.

File: external/libsodium/port/crypto_hash_mbedtls/crypto_hash_sha256_mbedtls.c [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+static void sha256_libsodium_to_mbedtls(mbedtls_sha256_context *mb_ctx, crypto_hash_sha256_state *ls_state)
+{
+	memcpy(mb_ctx->total, &ls_state->count, sizeof(mb_ctx->total));
+	memcpy(mb_ctx->state, ls_state->state, sizeof(mb_ctx->state));
+	memcpy(mb_ctx->buffer, ls_state->buf, sizeof(mb_ctx->buffer));
+	mb_ctx->is224 = 0;
+}
+
+int
+crypto_hash_sha256_init(crypto_hash_sha256_state *state)
```
Comment (xiarihf) [inline external/libsodium/port/crypto_hash_mbedtls/crypto_hash_sha256_mbedtls.c]: why do you separate definition of this and below functions?

File: external/libsodium/port/crypto_hash_mbedtls/crypto_hash_sha512_mbedtls.c [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+static void sha512_libsodium_to_mbedtls(mbedtls_sha512_context *mb_ctx, crypto_hash_sha512_state *ls_state)
+{
+	memcpy(mb_ctx->total, ls_state->count, sizeof(mb_ctx->total));
+	memcpy(mb_ctx->state, ls_state->state, sizeof(mb_ctx->state));
+	memcpy(mb_ctx->buffer, ls_state->buf, sizeof(mb_ctx->buffer));
+	mb_ctx->is384 = 0;
+}
+
+int
```
Comment (xiarihf) [inline external/libsodium/port/crypto_hash_mbedtls/crypto_hash_sha512_mbedtls.c]: No need to separate the function definition.

File: external/libsodium/port/randombytes_soft.c [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+{
+	if (time_v == (time_t)-1) {
+		time_v = time(NULL);
+		srand((unsigned int)time_v);
+	}
+
+	return (uint32_t)rand();
+}
+
+static void randombytes_soft_random_buf(void * const buf, const size_t size)
```
Comment (xiarihf) [inline external/libsodium/port/randombytes_soft.c]:
```suggestion
static void randombytes_soft_random_buf(void *const buf, const size_t size)
```

File: external/libsodium/port/randombytes_esp32.c [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "randombytes_default.h"
+#include "esp_system.h"
+
+static void randombytes_esp32_random_buf(void * const buf, const size_t size)
```
Comment (xiarihf) [inline external/libsodium/port/randombytes_esp32.c]:
```suggestion
static void randombytes_esp32_random_buf(void *const buf, const size_t size)
```

File: external/libsodium/port/sodium/version.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+SODIUM_EXPORT
+const char *sodium_version_string(void);
+
+SODIUM_EXPORT
+int         sodium_library_version_major(void);
```
Comment (xiarihf) [inline external/libsodium/port/sodium/version.h]:
```suggestion
int sodium_library_version_major(void);
```

File: external/libsodium/port/randombytes_soft.c [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+{
+	if (time_v == (time_t)-1) {
+		time_v = time(NULL);
+		srand((unsigned int)time_v);
+	}
+
+	uint8_t *p = (uint8_t *)buf;
+	for (size_t i = 0; i < size; i++) {
```
Comment (xiarihf) [inline external/libsodium/port/randombytes_soft.c]:
```suggestion
	size_t i;
	for (i = 0; i < size; i++) {
```
It would be a error in C89.

Comment (SRCN-LiQin) [reply external/libsodium/port/randombytes_soft.c]: fixed, thanks

File: external/libsodium/port/sodium/version.h (hunk 5 above)
Comment (SRCN-LiQin) [reply external/libsodium/port/sodium/version.h]: fixed, thanks

File: external/libsodium/port/randombytes_esp32.c (hunk 4 above)
Comment (SRCN-LiQin) [reply external/libsodium/port/randombytes_esp32.c]: fixed, thanks

File: external/libsodium/port/randombytes_soft.c (hunk 3 above)
Comment (SRCN-LiQin) [reply external/libsodium/port/randombytes_soft.c]: fixed, thanks

**Comment by seinfra** [comment]
Target : [f7190b6f10e1044a1d57c5fdb5d4bad0ea7debe6] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [f7190b6f10e1044a1d57c5fdb5d4bad0ea7debe6] - Code Rule Check (C++) OK.

File: external/libsodium/port/crypto_hash_mbedtls/crypto_hash_sha256_mbedtls.c (hunk 1 above)
Comment (SRCN-LiQin) [reply external/libsodium/port/crypto_hash_mbedtls/crypto_hash_sha256_mbedtls.c]: modified to one line. thanks

File: external/libsodium/port/crypto_hash_mbedtls/crypto_hash_sha512_mbedtls.c (hunk 2 above)
Comment (SRCN-LiQin) [reply external/libsodium/port/crypto_hash_mbedtls/crypto_hash_sha512_mbedtls.c]: modified to one line. thanks

**Comment by seinfra** [comment]
Target : [70e94c7cb32b5f00c43f245597ea07485f0bd343] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [70e94c7cb32b5f00c43f245597ea07485f0bd343] - Code Rule Check (C++) OK.

**Pull request merged**